File: Samples/Windows_Forms_C#/Form1.cs
using System;
using System.ComponentModel;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Linq;

using System.Text;
using System.Windows.Forms;
using System.Xml.Serialization;
using System.IO;

using System.Diagnostics;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Json;

using EntitySpaces.Interfaces;
using EntitySpaces.Core;
using EntitySpaces.DynamicQuery;
using EntitySpaces.Profiler;

using BusinessObjects;

namespace WindowsForms
{
    public partial class Form1 : Form
    {
        EmployeesCollection coll = new EmployeesCollection();
        bool acending = false;
        bool filtered = false;

        public Form1()
        {
            InitializeComponent();

            dataGridView1.AutoGenerateColumns = true;
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            coll = new EmployeesCollection();
            coll.LoadAll();

            _bindingSource.DataSource = coll;
        }

        private void btnSave_Click(object sender, EventArgs e)
        {
            try
            {
                coll.Save();
            }
            catch(Exception ex)
            {
                lblError.Text = ex.Message;
            }
        }

        private void btnSort_Click(object sender, EventArgs e)
        {
            if (acending)
            {
                acending = false;
                coll.Filter = coll.AsQueryable().OrderByDescending(d => d.LastName);
            }
            else
            {
                acending = true;
                coll.Filter = coll.AsQueryable().OrderBy(d => d.LastName);
            }
        }

        private void btnFilter_Click(object sender, EventArgs e)
        {
            if (filtered)
            {
                filtered = false;
                coll.Filter = null;
            }
            else
            {
                filtered = true;
                coll.Filter = coll.AsQueryable().Where(d => d.FirstName.Contains("a"));
            }
        }

        private void btnFilterSort_Click(object sender, EventArgs e)
        {
            if (acending)
            {
                acending = false;
                coll.Filter = coll.AsQueryable().Where(d => d.FirstName.Contains("a"));
            }
            else
            {
                acending = true;
                coll.Filter = coll.AsQueryable().Where(d => d.FirstName.Contains("a")).OrderBy(d => d.LastName);
            }
        }

        private void btnAdd_Click(object sender, EventArgs e)
        {
            Employees emp = coll.AddNew();
            emp.FirstName = "Joe";
            emp.LastName = "Smith";
        }

        private void btnDelete_Click(object sender, EventArgs e)
        {
            var selectedRows = dataGridView1.SelectedRows;

            List<Employees> list = new List<Employees>();
            foreach (DataGridViewRow item in selectedRows)
            {
                Employees emp = item.DataBoundItem as Employees;

                if (emp != null)
                {
                    list.Add(emp);
                }
            }

            foreach (Employees emp in list)
            {
                emp.MarkAsDeleted();
            }
        }

        private void btnChangeFirstName_Click(object sender, EventArgs e)
        {
            foreach (Employees emp in coll)
            {
                emp.FirstName = "Sally";
            }
        }

        private void btnRefresh_Click(object sender, EventArgs e)
        {
            btnAdd.Enabled = true;
            btnDelete.Enabled = true;
            btnSave.Enabled = true;

            coll = new EmployeesCollection();
            coll.LoadAll();

            _bindingSource.DataSource = coll;
        }

        private void btnDynamicQuery_Click(object sender, EventArgs e)
        {
            btnAdd.Enabled = false;
            btnDelete.Enabled = false;
            btnSave.Enabled = false;

            EmployeesQuery eq = new EmployeesQuery("eq");
            OrdersQuery oq = new OrdersQuery("oq");

            eq.Select(eq.EmployeeID, eq.LastName, eq.FirstName, oq.OrderID, oq.ShipCity);
            eq.InnerJoin(oq).On(eq.EmployeeID == oq.EmployeeID);

            coll = new EmployeesCollection();
            coll.Load(eq);

            _bindingSource.DataSource = coll;
        }

        private void btnBeginProfiling_Click(object sender, EventArgs e)
        {
            ProfilerListener.BeginProfiling("EntitySpaces.SqlClientProvider", 
                ProfilerListener.Channels.Channel_1);
        }

        private void btnEndProfiling_Click(object sender, EventArgs e)
        {
            ProfilerListener.EndProfiling("EntitySpaces.SqlClientProvider");
        }
    }
}

File: Samples/Windows_Forms_VB/Form1.vb
Imports WindowsFormsVB.BusinessObjects

Imports EntitySpaces.Interfaces
Imports EntitySpaces.Loader
Imports EntitySpaces.Profiler

Public Class Form1

    Private coll As New EmployeesCollection
    Private acending As Boolean
    Private filtered As Boolean


    Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        ' This only needs to be done once per the entire application
        esProviderFactory.Factory = New esDataProviderFactory()

        Try
            dataGridView1.AutoGenerateColumns = True
            coll.LoadAll()

            ' Bind to our EntitySpaces Collection
            _bindingSource.DataSource = coll

        Catch ex As Exception
            lblError.Text = ex.Message
        End Try

    End Sub

    Private Sub btnSave_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSave.Click
        Try
            coll.Save()
        Catch ex As Exception
            lblError.Text = ex.Message
        End Try
    End Sub

    Private Sub btnSort_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSort.Click

        If acending Then
            acending = False
            coll.Filter = coll.AsQueryable().OrderByDescending(Function(d) d.LastName)
        Else
            acending = True
            coll.Filter = coll.AsQueryable().OrderBy(Function(d) d.LastName)
        End If

    End Sub

    Private Sub btnFilter_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnFilter.Click

        If filtered Then
            filtered = False
            coll.Filter = Nothing
        Else
            filtered = True
            coll.Filter = coll.AsQueryable().Where(Function(d) d.FirstName.Contains("a"))
        End If

    End Sub

    Private Sub btnFilterSort_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnFilterSort.Click

        If acending Then
            acending = False
            coll.Filter = coll.AsQueryable().Where(Function(d) d.FirstName.Contains("a"))
        Else
            acending = True
            coll.Filter = coll.AsQueryable().Where(Function(d) d.FirstName.Contains("a")).OrderBy(Function(d) d.LastName)
        End If

    End Sub

    Private Sub btnAdd_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnAdd.Click

        Dim emp = coll.AddNew()
        emp.FirstName = "Joe"
        emp.LastName = "Smith"

    End Sub

    Private Sub btnDelete_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnDelete.Click

        Dim selectedRows = dataGridView1.SelectedRows

        Dim list As New List(Of Employees)()

        For Each item As DataGridViewRow In selectedRows
            Dim emp As Employees = TryCast(item.DataBoundItem, Employees)

            If emp IsNot Nothing Then
                list.Add(emp)
            End If
        Next

        For Each emp As Employees In list
            emp.MarkAsDeleted()
        Next

    End Sub

    Private Sub btnChangeFirstName_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnChangeFirstName.Click

        For Each emp As Employees In coll
            emp.FirstName = "Sally"
        Next

    End Sub

    Private Sub btnDynamicQuery_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnDynamicQuery.Click

        btnAdd.Enabled = False
        btnDelete.Enabled = False
        btnSave.Enabled = False

        Dim eq As New EmployeesQuery("eq")
        Dim oq As New OrdersQuery("oq")

        eq.[Select](eq.EmployeeID, eq.LastName, eq.FirstName, oq.OrderID, oq.ShipCity)
        eq.InnerJoin(oq).[On](eq.EmployeeID = oq.EmployeeID)

        coll = New EmployeesCollection()
        coll.Load(eq)

        _bindingSource.DataSource = coll

    End Sub

    Private Sub btnRefresh_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnRefresh.Click

        btnAdd.Enabled = True
        btnDelete.Enabled = True
        btnSave.Enabled = True

        coll = New EmployeesCollection()
        coll.LoadAll()

        _bindingSource.DataSource = coll

    End Sub

    Private Sub btnBeginProfiling_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnBeginProfiling.Click

        ProfilerListener.BeginProfiling("EntitySpaces.SqlClientProvider", ProfilerListener.Channels.Channel_1)

    End Sub

    Private Sub btnEndProfiling_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnEndProfiling.Click

        ProfilerListener.EndProfiling("EntitySpaces.SqlClientProvider")

    End Sub

End Class
